File: src/selector.ts
import { GatewayError } from "./errors.js";
import { HealthRegistry } from "./health.js";
import { AppConfig, CostTier, Endpoint, ModelCatalog, ModelConfig, OpenAIJsonBody, RouteConfig, SelectedModel } from "./types.js";

export function resolveModel(config: AppConfig, health: HealthRegistry, requestedModel: string, catalog?: ModelCatalog): SelectedModel {
  const directModel = catalog?.getModel(requestedModel) ?? (config.models[requestedModel] ? { name: requestedModel, config: config.models[requestedModel] } : undefined);
  if (directModel) {
    return directModel;
  }

  const route = config.routes[requestedModel];
  if (!route) {
    throw new GatewayError(`Unknown model or route: ${requestedModel}`, 404, "model_not_found");
  }

  return selectAutoModels(config, health, requestedModel, route, undefined, catalog)[0]!;
}

export function isAutoRoute(config: AppConfig, requestedModel: string): boolean {
  return config.routes[requestedModel] !== undefined;
}

export function resolveAutoCandidates(
  config: AppConfig,
  health: HealthRegistry,
  requestedModel: string,
  endpoint: Endpoint,
  body?: OpenAIJsonBody,
  catalog?: ModelCatalog,
): SelectedModel[] {
  const route = config.routes[requestedModel];
  if (!route) {
    throw new GatewayError(`Unknown model or route: ${requestedModel}`, 404, "model_not_found");
  }
  if (route.endpoint !== endpoint) {
    throw new GatewayError(`Route ${requestedModel} does not support endpoint ${endpoint}`, 400, "endpoint_mismatch");
  }
  return selectAutoModels(config, health, requestedModel, route, estimateRequestTokens(body), catalog);
}

function selectAutoModels(config: AppConfig, health: HealthRegistry, routeName: string, route: RouteConfig, estimatedTokens?: number, catalog?: ModelCatalog): SelectedModel[] {
  const entries = catalog?.getModelEntries() ?? Object.entries(config.models).map(([name, model]) => [name, model, undefined] as const);
  const candidates = entries
    .filter(([, model]) => model.endpoint === route.endpoint)
    .filter(([, model]) => !route.require_private || model.provider === "local")
    .filter(([modelName, model]) => !route.cost_tier || modelTier(modelName, model) === route.cost_tier)
    .filter(([, model]) => !estimatedTokens || !model.context_window || estimatedTokens <= model.context_window)
    .filter(([modelName]) => {
      const state = health.get(modelName);
      return state.healthy && !state.busy && latencyMatches(state.network_latency_ms, route.latency?.network_p95_ms_max) && latencyMatches(state.first_token_latency_ms, route.latency?.first_token_p95_ms_max);
    })
    .sort((a, b) => {
      const priceDelta = modelRank(a[1]) - modelRank(b[1]);
      if (priceDelta !== 0) return priceDelta;
      const orderDelta = modelOrder(config, a[0]) - modelOrder(config, b[0]);
      if (orderDelta !== 0) return orderDelta;
      return (health.get(a[0]).network_latency_ms ?? Number.MAX_SAFE_INTEGER) - (health.get(b[0]).network_latency_ms ?? Number.MAX_SAFE_INTEGER);
    })
    .map(([name, model, metadata]) => ({ name, config: model, metadata }));

  if (candidates.length === 0) {
    throw new GatewayError(`No healthy model matches route: ${routeName}`, 503, "no_healthy_model");
  }

  return candidates;
}

function latencyMatches(actual: number | undefined, max: number | undefined): boolean {
  return max === undefined || actual === undefined || actual <= max;
}

function modelRank(model: { provider: string; price_rank?: number }): number {
  if (model.price_rank !== undefined) return model.price_rank;
  if (model.provider === "local") return 0;
  if (model.provider === "deepseek") return 10;
  if (model.provider === "openrouter") return 20;
  return 100;
}

function modelOrder(config: AppConfig, modelName: string): number {
  return config.model_order[modelName] ?? Number.MAX_SAFE_INTEGER;
}

function modelTier(modelName: string, model: ModelConfig): CostTier {
  if (model.cost_tier) return model.cost_tier;
  if (model.provider === "local") return "free";
  if (model.provider === "deepseek") return "standard";
  if (modelName.endsWith("/premium")) return "premium";
  return "standard";
}

function estimateRequestTokens(body?: OpenAIJsonBody): number | undefined {
  if (!body) return undefined;
  const text = JSON.stringify(body);
  return Math.ceil(text.length / 4);
}

File: rust/selector.rs
use anyhow::{anyhow, Result};
use serde_json::Value;

use crate::{
    health::HealthRegistry,
    runtime_catalog::RuntimeCatalog,
    types::{AppConfig, Endpoint, ModelConfig, RouteConfig, SelectedModel},
};

pub fn resolve_model(
    config: &AppConfig,
    health: &HealthRegistry,
    requested_model: &str,
    catalog: Option<&RuntimeCatalog>,
) -> Result<SelectedModel> {
    if let Some(model) = catalog.and_then(|catalog| catalog.get_model(requested_model)) {
        return Ok(model);
    }
    if let Some(model) = config.models.get(requested_model) {
        return Ok(SelectedModel {
            name: requested_model.to_string(),
            config: model.clone(),
            metadata: None,
        });
    }
    let route = config
        .routes
        .get(requested_model)
        .ok_or_else(|| anyhow!("Unknown model or route: {requested_model}"))?;
    select_auto_models(config, health, requested_model, route, None, catalog)?
        .into_iter()
        .next()
        .ok_or_else(|| anyhow!("No healthy model matches route: {requested_model}"))
}

pub fn is_auto_route(config: &AppConfig, requested_model: &str) -> bool {
    config.routes.contains_key(requested_model)
}

pub fn resolve_auto_candidates(
    config: &AppConfig,
    health: &HealthRegistry,
    requested_model: &str,
    endpoint: Endpoint,
    body: Option<&Value>,
    catalog: Option<&RuntimeCatalog>,
) -> Result<Vec<SelectedModel>> {
    let route = config
        .routes
        .get(requested_model)
        .ok_or_else(|| anyhow!("Unknown model or route: {requested_model}"))?;
    if route.endpoint != endpoint {
        return Err(anyhow!(
            "Route {requested_model} does not support endpoint {:?}",
            endpoint
        ));
    }
    let mut candidates = select_auto_models(
        config,
        health,
        requested_model,
        route,
        estimate_request_tokens(body),
        catalog,
    )?;
    candidates.truncate(config.server.fallback_max_attempts.max(1));
    Ok(candidates)
}

fn select_auto_models(
    config: &AppConfig,
    health: &HealthRegistry,
    route_name: &str,
    route: &RouteConfig,
    estimated_tokens: Option<u64>,
    catalog: Option<&RuntimeCatalog>,
) -> Result<Vec<SelectedModel>> {
    let entries = if let Some(catalog) = catalog {
        catalog.get_model_entries()
    } else {
        config
            .models
            .iter()
            .map(|(name, model)| (name.clone(), model.clone(), None))
            .collect()
    };
    let mut candidates: Vec<_> = entries
        .into_iter()
        .filter(|(_, model, _)| model.endpoint == route.endpoint)
        .filter(|(_, model, _)| {
            !route.require_private.unwrap_or(false) || model.provider == "local"
        })
        .filter(|(name, model, _)| {
            route
                .cost_tier
                .as_ref()
                .is_none_or(|tier| model_tier(name, model) == *tier)
        })
        .filter(|(_, model, _)| {
            estimated_tokens
                .is_none_or(|tokens| model.context_window.is_none_or(|window| tokens <= window))
        })
        .filter(|(name, _, _)| {
            let state = health.get(name);
            state.healthy
                && !state.busy
                && latency_matches(
                    state.network_latency_ms,
                    route
                        .latency
                        .as_ref()
                        .and_then(|latency| latency.network_p95_ms_max),
                )
                && latency_matches(
                    state.first_token_latency_ms,
                    route
                        .latency
                        .as_ref()
                        .and_then(|latency| latency.first_token_p95_ms_max),
                )
        })
        .map(|(name, config, metadata)| SelectedModel {
            name,
            config,
            metadata,
        })
        .collect();

    candidates.sort_by(|a, b| {
        model_rank(&a.config)
            .cmp(&model_rank(&b.config))
            .then_with(|| model_order(config, &a.name).cmp(&model_order(config, &b.name)))
            .then_with(|| {
                health
                    .get(&a.name)
                    .network_latency_ms
                    .unwrap_or(u64::MAX)
                    .cmp(&health.get(&b.name).network_latency_ms.unwrap_or(u64::MAX))
            })
    });

    if candidates.is_empty() {
        return Err(anyhow!("No healthy model matches route: {route_name}"));
    }
    Ok(candidates)
}

fn latency_matches(actual: Option<u64>, max: Option<u64>) -> bool {
    max.is_none_or(|max| actual.is_none_or(|actual| actual <= max))
}

fn model_rank(model: &ModelConfig) -> i64 {
    if let Some(rank) = model.price_rank {
        return rank;
    }
    match model.provider.as_str() {
        "local" => 0,
        "deepseek" => 10,
        "openrouter" => 20,
        _ => 100,
    }
}

fn model_order(config: &AppConfig, model_name: &str) -> usize {
    config
        .model_order
        .get(model_name)
        .copied()
        .unwrap_or(usize::MAX)
}

fn model_tier(model_name: &str, model: &ModelConfig) -> String {
    if let Some(tier) = &model.cost_tier {
        return tier.clone();
    }
    if model.provider == "local" {
        return "free".to_string();
    }
    if model.provider == "deepseek" {
        return "standard".to_string();
    }
    if model_name.ends_with("/premium") {
        return "premium".to_string();
    }
    "standard".to_string()
}

fn estimate_request_tokens(body: Option<&Value>) -> Option<u64> {
    let body = body?;
    let text = serde_json::to_string(body).ok()?;
    Some((text.len() as u64).div_ceil(4))
}
